### ISSUE TITLE
Global Auto-Type is slow

### ISSUE THREAD

**Issue by loilo** [keeweb/keeweb#347]
This may come around as unconstructive critics but it's rather meant as a reality-check to compare with other users:

When running the global Auto-Type shortcut ^⌥T resp. Alt + Shift + T it takes KeeWeb a pretty long time (compared to KeePass) to fill in the right password – usually several seconds.

This is the case on my Windows machine as well as on my MacBook Pro. They're both not the slowest computers so I don't think it's a hardware issue.

Is this a generally known problem or am I a strange exception?

**Comment by antelle** [comment]
Are you using auto-type obfuscation?

**Comment by Loilo** [comment]
I'm not using obfuscation.

Here are the logs:

> [auto-type] Auto type event null
> [auto-type] Get window title
> [launcher] spawn osascript: 0, 423ms
> Opera
> [launcher] spawn osascript: 0, 3254ms
> This is the replaced window title
> [auto-type] Window title This is the replaced window title undefined
> [auto-type] Hide window
> [auto-type] Start {USERNAME}{TAB}{PASSWORD}{ENTER}
> [auto-type] Parsed [op:USERNAME,op:TAB,op:PASSWORD,op:ENTER]
> [auto-type] Resolved [****************,key:tab,[*,*,*,*,*,*,*,*,*,*],key:enter]
> [launcher] spawn /Users/myusername/Library/Application Support/KeeWeb/helper/darwin/KeeWebHelper: 0, 1863ms
> [auto-type] Complete 1868ms


It's noticeable that this line's duration

> [launcher] spawn osascript: 0, 3254ms

seems to reduce with multiple usages. First usage printed 3254ms, second try was about 1800ms and from the third try on it's around 175ms.

I'll try again later and check if the script will eventually take a long time to run again.

**Comment by antelle** [comment]
Yes, first init time may be very slow. We are getting urls from accessibility API, which takes some time to init... Here's AppScript which we use: https://github.com/keeweb/keeweb/blob/master/app/scripts/auto-type/helper/auto-type-helper-darwin.js#L5

**Comment by Loilo** [comment]
I couldn't figure out if it's actually faster in Chrome and Safari. Those were the 3rd and 4th time I tried and thus all went pretty fast anyway.

I'd have guessed that initialization could be the sticking point here, but to be honest that's a rather strong one. Typically I want to use Auto Type once in a while and between to usages the script seems to have cooled down.

**Comment by antelle** [comment]
The problem is, it's something we don't control. I'm not sure if moving logic to helper app will help.

**Comment by grinsted** [comment]
Im not sure if this is feasible or would help at all. (i have zero experience in node or applescript). - But would it be possible/help to use osacompile on the scripts on first run, and pass the strings as commandline args?

**Comment by antelle** [comment]
I want to check first, if implementing it in Objective C helps. We have a helper which has native code inside, so it should be faster, but I'm not sure; this may be API initialization time, then we'll not  be able to speedup.

**Comment by MichalKarnicki** [comment]
For the record, I haven't used much of KeeWeb's auto type yet, but I sure intend to. I have a suggestion though.

Would it be possible to determine appropriate password picking for auto type based on window title instead the of the url? On Mac, you can see the window title (which is normally not visible) in the following menu:
Menu -> Window

I found this, may be helpful. [link](https://github.com/octalmage/active-window)

I recall KeePass had an auto type configuration feature that enabled the user to customize and/or provide specific window titles in a given KeePass entry (such as, Auto-Type-Window: [website title here]). So you could use that, not only the url.

Sorry if my comment is not on point, I'm new to KeeWeb :)

**Comment by antelle** [comment]
KeeWeb is using both window URL and title for auto-type detection.

**Comment by Metamist** [comment]
I may be late to the party, but I noticed that auto type is very slow in Chrome, but near instant in Firefox. Looking at the logs tells me that KeeWebHelper.exe takes about 6 seconds to get the window title / address,
```
[DEBUG] 2017-01-03T13:22:13.938Z [auto-type] Auto type event 
[DEBUG] 2017-01-03T13:22:13.943Z [auto-type] Get window title
[INFO ] 2017-01-03T13:22:19.662Z [launcher] spawn D:\Program\KeeWeb\resources\app\helper\win32\KeeWebHelper.exe: 0, 5719ms
Facepunch - Google Chrome
https://facepunch.com/forum.php
[DEBUG] 2017-01-03T13:22:19.662Z [auto-type] Window title Facepunch - Google Chrome https://facepunch.com/forum.php
[DEBUG] 2017-01-03T13:22:19.662Z [auto-type] Hide window
...
```

The same website, except in Firefox, which only takes like 500ms:
```
[DEBUG] 2017-01-03T13:26:11.240Z [auto-type] Auto type event 
[DEBUG] 2017-01-03T13:26:11.244Z [auto-type] Get window title
[INFO ] 2017-01-03T13:26:11.763Z [launcher] spawn D:\Program\KeeWeb\resources\app\helper\win32\KeeWebHelper.exe: 0, 520ms
Facepunch - Mozilla Firefox
https://facepunch.com/forum.php
[DEBUG] 2017-01-03T13:26:11.764Z [auto-type] Window title Facepunch - Mozilla Firefox https://facepunch.com/forum.php
[DEBUG] 2017-01-03T13:26:11.764Z [auto-type] Hide window
...
```

I'm on Windows 8.1, can't speak for other Windows versions except 10, I'll test that out soon and edit this post.

**Comment by antelle** [comment]
@Metamist, what happens if you trigger auto-type in Chrome twice? First-time initialization may take some time...

**Comment by Metamist** [comment]
@antelle Doesn't matter how many times I trigger it, it always takes about 6 seconds for it to process and either input the password (if there is only one) or show me the list of users.

**Comment by antelle** [comment]
that's weird...

**Comment by Metamist** [comment]
It seems to be much faster using my laptop which is running windows 10, currently takes around 1.6 seconds but still not as fast as Firefox or Edge. Is there any way I can debug why it takes such a long time on my desktop PC?

**Comment by dploeger** [comment]
Just wanted to ping this, because it's still valid on Chrome for macOS. Safari seems to be quicker. Really strange.

**Comment by dploeger** [comment]
What I did notice is, that Autotype really only starts when you release meta keys. So when you hit Option-Enter, Autotype will only start when you have released both keys. I tend to keep hold on the meta key (somehow). Maybe that confuses people.

**Comment by antelle** [comment]
This may be fixed in v1.17

**Comment by antelle** [comment]
Since it's an old issue that is very likely to be fixed in the latest version, closing it. Please let me know if it still repeats.
